kep/640-error-pods-rescheduling/README.md

```diff
@@ -0,0 +1,166 @@
+* Implement pod rescheduling plugins in PreFilter extension point to prefilter nodes that the pod has already scheduled.
+* Pod rescheduling plugin in this proposal has no conflict with default scheduler's plugins.
+
+### Non-Goals
+
+* Reschedule pods controlled by daemonset or other workload resources which don't exist in default k8s controller.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+A pod submitted to the Kubernetes cluster initially operates smoothly, but over time, as the memory on the node hosting the pod becomes insufficient and the pod's memory requirements grow, it eventually gets CrashLoopBackOff due to out-of-memory (OOM) issues. 
+Consequently, we have to manually relocate the pod to other nodes. But this pod may schedule to the same node with the default scheduler.
+
+#### Story 2
+
+After a pod is submitted to the Kubernetes cluster and successfully scheduled, it may encounter startup issues due to mismatches in hardware resources on the node, 
+such as NUMA or device-related information. In such cases, we need to manually reschedule the pod to a different node. But this pod may schedule to the same node with the default scheduler.
+
+## Design Details
+
+Our scheme is to design two components as outlined in the diagram and described below: Rescheduler-controller and Podrescheduling. 
+Users can configure the `scheduling-retries` (maximum rescheduling retry count) in the metadata annotations of workload resources, such as Deployment and StatefulSet. 
+Rescheduler-controller is responsible for monitoring pods with abnormal states in the cluster. For each abnormal pod, 
+the rescheduler-controller checks the current scheduling count (defined as `currentReschedulingTimes`) recorded in the metadata annotations of its associated workload resource. 
+If the current scheduling count does not exceed the user-configured `scheduling-retries` (maximum rescheduling retry count), 
+it increments the current scheduling count by 1 and records the current node where the pod is located in the scheduled nodes list (eg: defined as `deployScheduledHosts`). 
+Then, rescheduler-controller deletes the abnormal pod, and this pod's associated workload resource will create a new pod automatically. 
+Rescheduler-controller registers a MutateWebHook in the kube-apiserver, this MutateWebHook will patch scheduled nodes list info to this newly created pod's annotations (defined as `sigs.k8s.io/scheduled-hosts`) when this newly created pod comes to kube-apiserver admission point.  
+Podrescheduling is the scheduler-plugins that implement the PreFilter extension point of kube-scheduler to prefilter out the scheduled nodes list recorded in this pod annotation and schedule this pod to other nodes.
+
+![design](./design.png)
+
+### Rescheduler-controller
+
+The rescheduler-controller is comprised of two components, as illustrated in the above diagram: ListFunc and MutateWebHook.
```

Comment:
Maybe I am just peaky but why `ListFunc` component is also responsible for deleting pods?
It should be either `ListAndDeleteFunc` components or it should be divided into `ListFunc` and `DeleteFunc` components.

Reply:
@PiotrProkop Thanks for your review! I will update `ListFunc` to `ListAndDeleteFunc`.